### PR TITLE
Add optional TLS Support

### PR DESCRIPTION
**Pull request by stevendanna**
This adds TLS support to the memcache client. Users opt-in to a TLS
connection by using a URL in the form of:

   memcache+tls://HOST:PORT

TLS was added to memcached in version 1.5.13.

TLS options are available as query parameters, in line with the
previous tcp options. The available options are:

verify_mode = peer | none (default peer)
ca_path = PATH
key_path = PATH
cert_path = PATH

This adds the openssl crate as a dependency.

I've split this into 2 commits, the first commit is a small refactoring
that made it a bit easier to add a new transport.

If this PR is of interest to you, please let me know and I can spend a bit
more time adding tests into your existing test setup.

### CODE REVIEW

**Comment by aisk** [comment]
Great job, I'll try to check this tomorrow, thanks!

File: Cargo.toml [hunk 1]
```diff
@@ -13,3 +13,6 @@ byteorder = "1"
 url = "1"
 rand = "0.7"
 enum_dispatch = "0.1"
+# If we don't want to depend on the system openssl, we could do:
+# openssl = {version = "*", features = ["vendored"]}
+openssl = "*"
```
Comment (aisk) [inline Cargo.toml]: openssl's compilation is alway trouble some, can we mark it as an optional depencence?

File: src/connection.rs [hunk 2]
```diff
@@ -11,104 +11,183 @@ use error::MemcacheError;
 use protocol::{AsciiProtocol, BinaryProtocol, Protocol};
 use stream::Stream;
 use stream::UdpStream;
+use openssl::ssl::{SslConnector, SslMethod, SslFiletype, SslVerifyMode};
```
Comment (aisk) [inline src/connection.rs]: `use` statement shoud be ordered. May be a `cargo fmt` should resolve these.

File: src/connection.rs [hunk 3]
```diff
@@ -11,104 +11,183 @@ use error::MemcacheError;
...
 
 /// a connection to the memcached server
 pub struct Connection {
     pub protocol: Protocol,
     pub url: String,
 }
 
-impl Connection {
-    pub(crate) fn connect(url: &Url) -> Result<Self, MemcacheError> {
+enum Transport {
```
Comment (aisk) [inline src/connection.rs]: Words like `TCP` / `UDP` / `TLS` should using lower case with first char upper case. Maybe all upper case is right, but for now we should keep these like other places in this project.

**Comment by aisk** [comment]
Sorry late for review, I think your PR is cool with small issue in comment. And I think there should be a test case for TLS to make sure we do not break TLS codes in future.

**Comment by stevendanna** [comment]
@aisk Thanks for the review! I'll try to get back to this sometime this week.

**Comment by stevendanna** [comment]
Just wanted to know I haven't forgotten about this, just had something come up. Thanks again for the review.

**Comment by stevendanna** [comment]
Apologies for the rebase, it took a bit of experimentation to get the tests working in CI.

- I've added some basic connection tests for the TLS code.  This required some changes to the travis CI setup as the version of memcache available in travis did not support TLS by default.  To support these tests, I've checked in some test certificates, these certs expire in 10 years.

- The TLS code is now an optional feature.  I've made it default enabled for now, but let me know if you'd prefer it be disabled by default (should be a small Cargo.toml change)

- When working on getting the tests working in CI, I did see this failure once: https://travis-ci.org/aisk/rust-memcache/jobs/625698466 but haven't been able to reproduce it locally.

**Comment by aisk** [comment]
I'll review it this after work, thanks @stevendanna !

**Comment by aisk** [comment]
Epic work! thanks @stevendanna , I saw a lots refactor works for the CI setup and the connecting codes, realy cool.

I think it's fine to enable TLS by default, cause if someone can't compile SSL, he has option to disable it : https://doc.rust-lang.org/cargo/reference/manifest.html#the-features-section . We can keep this behavior until someone complain this.

**Pull request merged**